Updates to ParmEst documentation

slilonfe5 opened the pull request:
## Summary/Motivation:
The current documentation of ParmEst does not provide a detailed and step-by-step guide on using ParmEst to perform model-building tasks such as parameter estimation and covariance matrix calculation.

## Changes proposed in this PR:
- Add a step-by-step guide on how to create the Experiment class and its list using the simple Rooney-Biegler paper example
- Link the parameter estimation page to the covariance matrix estimation page to refer users to creating the Experiment class and its list.

### Legal Acknowledgement

By contributing to this software project, I have read the [contribution guide](https://pyomo.readthedocs.io/en/stable/contribution_guide.html) and agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the BSD license.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

slilonfe5 committed:
Updated covariance.rst file

slilonfe5 committed:
Minor updates to covariance.rst file

slilonfe5 committed:
Merge branch 'main' into parmest-doc-update

blnicho committed:
Merge branch 'main' into parmest-doc-update

slilonfe5 committed:
Updated driver.rst and covariance.rst files

slilonfe5 committed:
Merge branch 'parmest-doc-update' of https://github.com/slilonfe5/pyomo into parmest-doc-update

slilonfe5 committed:
Updated driver.rst and installation.rst files

slilonfe5 committed:
Updated driver.rst

slilonfe5 committed:
Updated driver.rst, overview.rst, and covariance.rst files

slilonfe5 committed:
Minor update to covariance.rst

slilonfe5 committed:
Update overview.rst

blnicho committed:
Merge branch 'main' into parmest-doc-update

blnicho committed:
Merge branch 'main' into parmest-doc-update

blnicho commented on doc/OnlineDocs/explanation/analysis/parmest/driver.rst:
```diff
@@ -23,8 +23,170 @@ modified within parmest to match the required specifications.
+* ``m.experiment_outputs`` which defines experiment output (Pyomo `Param`, `Var`, or `Expression`)
+  and their associated data values (float, int).
+* ``m.unknown_parameters`` which defines the mutable parameters or variables (Pyomo `Param` or `Var`)
+  to estimate along with their component unique identifier (Pyomo `ComponentUID`).
```
You should include `m.measurement_error` in this list

slilonfe5 replied:
Thanks for catching this.

blnicho commented on doc/OnlineDocs/explanation/analysis/parmest/driver.rst:
```diff
@@ -23,8 +23,170 @@ modified within parmest to match the required specifications.
+parmest requires that the user creates an :class:`~pyomo.contrib.parmest.experiment.Experiment` class that
+contains the annotated Pyomo model with labeled experiment outputs, unknown parameters, and measurement errors.
+
+A labeled Pyomo model ``m`` has the following additional suffixes (Pyomo `Suffix`):
```
```suggestion
parmest requires that the user create an :class:`~pyomo.contrib.parmest.experiment.Experiment` class that
builds an annotated Pyomo model denoting experiment outputs, unknown parameters, and measurement errors using Pyomo `Suffix` components.
```

blnicho commented on doc/OnlineDocs/explanation/analysis/parmest/driver.rst:
```diff
@@ -23,8 +23,170 @@ modified within parmest to match the required specifications.
+* ``m.experiment_outputs`` which defines experiment output (Pyomo `Param`, `Var`, or `Expression`)
+  and their associated data values (float, int).
+* ``m.unknown_parameters`` which defines the mutable parameters or variables (Pyomo `Param` or `Var`)
+  to estimate along with their component unique identifier (Pyomo `ComponentUID`).
```
Rewording suggestion:
```suggestion
* ``m.experiment_outputs`` maps the experiment output, or measurement, terms in the model (Pyomo `Param`, `Var`, or `Expression`) to their associated data values (float, int).
* ``m.unknown_parameters`` maps the model parameters to estimate (Pyomo `Param` or `Var`)
  to their component unique identifier (Pyomo `ComponentUID`) which is used to identify equivalent parameters across multiple experiments.
```

blnicho commented on doc/OnlineDocs/explanation/analysis/parmest/driver.rst:
```diff
@@ -23,8 +23,170 @@ modified within parmest to match the required specifications.
+    ...         self.label_model()
+    ...         self.finalize_model()
+    ...
+    ...         return self.model
```
Please replace this with a literal include block that pulls the experiment class directly from the py file in the repo. Duplicating the code here makes it too easy for things to become out of sync.

slilonfe5 replied:
Got it. Thanks.

blnicho commented on doc/OnlineDocs/explanation/analysis/parmest/driver.rst:
```diff
@@ -23,8 +23,170 @@ modified within parmest to match the required specifications.
+    ...         return self.model
+
+
+Step 2: Load the Data and Create a List of the Model's Experiment Class
```
Suggested rewording:
```suggestion
Step 2: Load the Data and Create a List Experiments
```

blnicho commented on doc/OnlineDocs/explanation/analysis/parmest/driver.rst:
```diff
@@ -88,66 +237,47 @@ Optionally, solver options can be supplied, e.g.,
+The second argument is an optional argument if the objective function has already been included in the
+Pyomo model, which defines the optimization objective function to use in parameter estimation. However, if the
+objective function has not been included in the Pyomo model, like the one in the :ref:`ExperimentClass` Section
+above, the user is required to supply the second argument.
```
Suggested rewording:
```suggestion
The second argument, ``obj_function``, is used to specify the objective function to use for parameter estimation if the user has not manually defined an objective function in their ``Experiment`` class.
```

blnicho commented on doc/OnlineDocs/explanation/analysis/parmest/driver.rst:
```diff
@@ -88,66 +237,47 @@ Optionally, solver options can be supplied, e.g.,
+objective function has not been included in the Pyomo model, like the one in the :ref:`ExperimentClass` Section
+above, the user is required to supply the second argument.
 
 If no objective function is specified, the Pyomo model is used "as is" and
```
```suggestion
If ``obj_function`` is not specified, the Pyomo model is used "as is" and
```

slilonfe5 committed:
Update doc/OnlineDocs/explanation/analysis/parmest/covariance.rst

Co-authored-by: Bethany Nicholson <blnicho@users.noreply.github.com>

slilonfe5 committed:
Update doc/OnlineDocs/explanation/analysis/parmest/driver.rst

Co-authored-by: Bethany Nicholson <blnicho@users.noreply.github.com>

slilonfe5 committed:
Update doc/OnlineDocs/explanation/analysis/parmest/driver.rst

Co-authored-by: Bethany Nicholson <blnicho@users.noreply.github.com>

slilonfe5 committed:
Update doc/OnlineDocs/explanation/analysis/parmest/driver.rst

Co-authored-by: Bethany Nicholson <blnicho@users.noreply.github.com>

slilonfe5 committed:
Update doc/OnlineDocs/explanation/analysis/parmest/driver.rst

Co-authored-by: Bethany Nicholson <blnicho@users.noreply.github.com>

slilonfe5 committed:
Update doc/OnlineDocs/explanation/analysis/parmest/driver.rst

Co-authored-by: Bethany Nicholson <blnicho@users.noreply.github.com>

slilonfe5 committed:
Update doc/OnlineDocs/explanation/analysis/parmest/driver.rst

Co-authored-by: Bethany Nicholson <blnicho@users.noreply.github.com>

slilonfe5 committed:
Update doc/OnlineDocs/explanation/analysis/parmest/driver.rst

Co-authored-by: Bethany Nicholson <blnicho@users.noreply.github.com>

slilonfe5 committed:
Update doc/OnlineDocs/explanation/analysis/parmest/driver.rst

Co-authored-by: Bethany Nicholson <blnicho@users.noreply.github.com>

slilonfe5 committed:
Implemented Bethany's feedback

slilonfe5 committed:
Merge branch 'parmest-doc-update' of https://github.com/slilonfe5/pyomo into parmest-doc-update

slilonfe5 committed:
Updated driver.rst and installation.rst files

slilonfe5 committed:
Updated driver.rst

slilonfe5 committed:
Updated driver.rst file

slilonfe5 committed:
Updated driver.rst file

blnicho committed:
Merge branch 'main' into parmest-doc-update

blnicho committed:
Merge branch 'main' into parmest-doc-update

blnicho commented on doc/OnlineDocs/explanation/analysis/parmest/driver.rst:
```diff
@@ -105,74 +123,20 @@ The experiment class has one required method:
 mathematical model outlined in the introduction section of this Quick Start.
 
-.. doctest::
+.. literalinclude:: ../../../pyomo/contrib/parmest/examples/rooney_biegler/rooney_biegler.py
```
This isn't rendering in the documentation. I think the path is slightly off. 
```suggestion
.. literalinclude:: ../../pyomo/contrib/parmest/examples/rooney_biegler/rooney_biegler.py
```

blnicho committed:
Revise Parmest Quick Start Guide content

Updated the Parmest Quick Start Guide to clarify the usage of the Experiment class and objective functions for parameter estimation.

slilonfe5 commented:
> @slilonfe5 I went ahead and made some significant edits in `driver.rst` to make the text more concise. Once the docs are done building I will verify that everything looks right and approve this PR.

@blnicho sounds great. Thank you

blnicho committed:
Fix path for literalinclude in driver.rst
